### PR TITLE
[ADM-783][Frontend] Keep user's input for board/pipeline/source control of config page

### PR DESCRIPTION
**Pull request by gabralia**
## Summary

Reference: https://dorametrics.atlassian.net/browse/ADM-783

1. refactor config for board/pipeline/source 
2. fix board/pipeline/source input change strategy
3. update validation strategy

## Before

https://github.com/au-heartbeat/Heartbeat/assets/17806850/c33a2f99-7cda-4c0f-8c5a-af93cd12a825

## After

board/pipeline/source control of config page will keep user's input

### CODE REVIEW

File: frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx [hunk 1]
```diff
@@ -138,6 +138,24 @@ describe('PipelineTool', () => {
     expect(getByText(TOKEN_ERROR_MESSAGE[1])).toHaveStyle(ERROR_MESSAGE_COLOR);
   });
 
+  it('should not show error message when field does not trigger any event given an empty value', () => {
+    setup();
+
+    screen.getByTestId('pipelineToolTextField').querySelector('input') as HTMLInputElement;
```
Comment (guzhongren) [inline frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx:144]: Avoid direct Node access. Prefer using the methods from Testing Library.

File: frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx [hunk 2]
```diff
@@ -138,6 +138,24 @@ describe('PipelineTool', () => {
...
+
+    expect(screen.queryByText(TOKEN_ERROR_MESSAGE[1])).not.toBeInTheDocument();
+  });
+
+  it('should show error message when focus on field given an empty value', () => {
+    setup();
+
+    const tokenInput = screen.getByTestId('pipelineToolTextField').querySelector('input') as HTMLInputElement;
```
Comment (guzhongren) [inline frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx:152]: Avoid direct Node access. Prefer using the methods from Testing Library.

File: frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx [hunk 3]
```diff
@@ -138,6 +138,24 @@ describe('PipelineTool', () => {
...
+    fireEvent.focus(tokenInput, { target: { value: '' } });
```
Comment (guzhongren) [inline frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx:153]: use userEvent, not fireEvent

File: frontend/__tests__/containers/ConfigStep/SourceControl.test.tsx [hunk 4]
```diff
@@ -125,6 +125,24 @@ describe('SourceControl', () => {
...
+    screen.getByTestId('sourceControlTextField').querySelector('input') as HTMLInputElement;
+
+    expect(screen.queryByText(TOKEN_ERROR_MESSAGE[1])).not.toBeInTheDocument();
+  });
+
+  it('should show error message when focus on field given an empty value', () => {
+    setup();
+
+    const tokenInput = screen.getByTestId('sourceControlTextField').querySelector('input') as HTMLInputElement;
+    fireEvent.focus(tokenInput, { target: { value: '' } });
```
Comment (guzhongren) [inline frontend/__tests__/containers/ConfigStep/SourceControl.test.tsx:140]: same issue, see the more issue in the PR, raised by `Codacy Production / Codacy Static Code Analysis`

**Comment by guzhongren** [comment]
remember following our PR template, don't customize it

File: frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx (hunk 3 above)
Comment (gabralia) [reply frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx:153]: `focus` event is not in `userEvent`

File: frontend/src/containers/ConfigStep/Board/index.tsx [hunk 5]
```diff
@@ -123,33 +48,30 @@ export const Board = () => {
...
-              }}
-              error={!field.isRequired || !field.isValid}
-              type={field.key === 'Token' ? 'password' : 'text'}
-              helperText={field.errorMessage}
-              sx={{ gridColumn: `span ${field.col}` }}
+              value={value}
+              onFocus={() => validateField(key)}
+              onChange={(e) => updateField(key, e.target.value)}
+              error={!!validatedError || !!verifiedError}
+              type={key === 'Token' ? 'password' : 'text'}
```
Comment (SimonTal) [inline frontend/src/containers/ConfigStep/Board/index.tsx:60]: extract Token as enum or const?

File: frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx [hunk 6]
```diff
@@ -138,6 +138,21 @@ describe('PipelineTool', () => {
...
+  it('should not show error message when field does not trigger any event given an empty value', () => {
+    setup();
+
+    expect(screen.queryByText(TOKEN_ERROR_MESSAGE[1])).not.toBeInTheDocument();
+  });
+
+  it('should show error message when focus on field given an empty value', () => {
+    setup();
+
+    fireEvent.focus(screen.getByLabelText('input Token'));
```
[the remark this answers is not in the document]
Comment (gabralia) [reply frontend/__tests__/containers/ConfigStep/PipelineTool.test.tsx:150]: focus event is not in userEvent

**Review by SimonTal — APPROVED** [review]
Approved

**Pull request merged**